# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/datasets/music_project_en.csv')

Muestra las 10 primeras filas de la tabla:

In [4]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [7]:
# Obtener la información general sobre nuestros datos
df.describe(include="all")

userID  Track     artist  genre       City        time     Day
count      65079  63736      57512  63881        65079     65079   65079
unique     41748  39666      37806    268            2     20392       3
top     A8AE9169  Brand  Kartvelli    pop  Springfield  21:51:22  Friday
freq          76    136        136   8850        45360        14   23149

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
En la tabla df, los datos incluyen:

userID: Identificadores únicos de usuarios (tipo object).
Track: Nombres de canciones reproducidas (tipo object).
artist: Nombre del artista de la canción (tipo object).
genre: Género musical de la canción (tipo object).
City: Ciudad desde donde se escuchó la canción (tipo object).
time: Hora de reproducción (tipo object).
Day: Día de la semana en que se reprodujo la canción (tipo object).

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
El DataFrame tiene 65,079 filas, lo que indica un conjunto de datos relativamente grande. Sin embargo, hay valores ausentes en algunas columnas, como:

Track: 63,736 valores no nulos (hay datos faltantes).
artist: 57,512 valores no nulos (algunos artistas no están registrados).
genre: 63,881 valores no nulos (algunas canciones no tienen género especificado).

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
Sí, hay varios problemas a considerar:

Valores ausentes:

artist, Track y genre tienen datos faltantes, lo que puede afectar los análisis si no se manejan correctamente.
Inconsistencias en los nombres de las columnas:

Algunas columnas tienen nombres en mayúsculas (City, Day) y otras en minúsculas (artist, genre), lo que puede generar problemas al acceder a ellas.
Se recomienda estandarizar los nombres, por ejemplo, con df.columns = df.columns.str.lower().
Posibles datos incorrectos o inconsistentes:

La columna time tiene formato object, lo que indica que no se ha convertido en un formato de tiempo (datetime).

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [11]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [12]:
# Bucle en los encabezados poniendo todo en minúsculas
df.columns = df.columns.str.lower().str.replace(" ", "_")

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [19]:
# Bucle en los encabezados eliminando los espacios
df.columns = df.columns.str.strip()
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace("_", "")

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [20]:
# Cambiar el nombre de la columna "userid"
df.rename(columns={"userid": "user_id"}, inplace=True)

Comprueba el resultado. Muestra los encabezados una vez más:

In [21]:
# Comprobar el resultado: la lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [22]:
# Calcular el número de valores ausentes
missing_values = df.isnull().sum()
print(missing_values)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [23]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
df = df.fillna("unknown")

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [25]:
# Contar valores ausentes
missing_values = df.isnull().sum()
print(missing_values)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [26]:
# Contar duplicados explícitos
num_duplicados = df.duplicated().sum()
print(f"Número de filas duplicadas: {num_duplicados}")

Número de filas duplicadas: 3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [27]:
# Eliminar duplicados explícitos
# Eliminar las filas duplicadas y mantener solo la primera aparición
df = df.drop_duplicates()

# Mostrar el número de filas después de la eliminación
print(f"Número de filas después de eliminar duplicados: {df.shape[0]}")

Número de filas después de eliminar duplicados: 61253


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [28]:
# Comprobar de nuevo si hay duplicados
num_duplicados = df.duplicated().sum()

# Mostrar el resultado
print(f"Número de filas duplicadas después de la limpieza: {num_duplicados}")

Número de filas duplicadas después de la limpieza: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [29]:
# Inspeccionar los nombres de géneros únicos
# Obtener los valores únicos de la columna "genre", ordenados alfabéticamente
unique_genres = df["genre"].dropna().unique()
unique_genres.sort()

# Mostrar los géneros únicos
print(unique_genres)

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [30]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, wrong_genres, correct_genre):
    """
    Reemplaza valores en la columna 'genre' de df, sustituyendo
    los valores en wrong_genres por correct_genre.
    
    Parámetros:
    df (DataFrame): El DataFrame donde se realizará la corrección.
    wrong_genres (list): Lista de valores incorrectos a reemplazar.
    correct_genre (str): El valor correcto con el que se reemplazarán.
    """
    for genre in wrong_genres:
        df["genre"] = df["genre"].replace(genre, correct_genre)


Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [33]:
# Eliminar duplicados implícitos
# Eliminar duplicados implícitos en la columna "genre"
wrong_genres = ["hip", "hop", "hip-hop"]  # Lista de géneros incorrectos
correct_genre = "hiphop"  # Nombre correcto

# Aplicar la función para corregir los géneros duplicados
replace_wrong_genres(df, wrong_genres, correct_genre)

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [34]:

# Comprobar si los valores han sido reemplazados correctamente
print("Géneros únicos después de la corrección:")
print(df["genre"].unique())

Géneros únicos después de la corrección:
['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
### **📌 Observaciones sobre los duplicados y su eliminación**

1. **Análisis de los duplicados:**
   - Se encontraron dos tipos de duplicados:
     - **Duplicados explícitos:** Filas idénticas que aparecían más de una vez.
     - **Duplicados implícitos:** Variaciones en los valores, especialmente en la columna `genre` (por ejemplo, `"hip"`, `"hop"`, `"hip-hop"` referían al mismo género).

2. **Enfoque para eliminarlos:**
   - Para **duplicados explícitos**, se utilizó `df.drop_duplicates()`, eliminando filas repetidas mientras se conservaba la primera ocurrencia.
   - Para **duplicados implícitos**, se creó una función (`replace_wrong_genres()`) que reemplazó nombres alternativos en `genre` por una versión estandarizada (`hiphop`).

3. **Resultados obtenidos:**
   - Se eliminaron todas las filas duplicadas explícitas, reduciendo el tamaño del conjunto de datos sin pérdida de información relevante.
   - Los valores en la columna `genre` quedaron normalizados, evitando inconsistencias que podrían afectar futuros análisis.
   - Al comprobar con `df.duplicated().sum()` y `df["genre"].unique()`, se confirmó que los duplicados fueron eliminados correctamente.

 **Este proceso mejoró la limpieza y calidad de los datos, asegurando una mejor interpretación en análisis futuros.**

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [35]:

# Contar el número de canciones reproducidas en cada ciudad
tracks_per_city = df.groupby("city")["user_id"].count()

# Mostrar el resultado
print(tracks_per_city)


city
Shelbyville    18512
Springfield    42741
Name: user_id, dtype: int64


Son de tipo int64, de la lista 'user_id', se puede observar en la columna city dos nombres de ciudades 'Shelbyville' con 18512 datos y la ciudad de 'Springfield' con 42741 datos o reproducciones.
 Springfield tiene aproximadamente 2.3 veces más reproducciones que Shelbyville, lo que sugiere que los usuarios de Springfield son más activos en la plataforma.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [36]:
# Calcular las canciones reproducidas en cada uno de los tres días

tracks_per_day = df[df["day"].isin(["Monday", "Wednesday", "Friday"])].groupby("day")["user_id"].count()

# Mostrar el resultado
print(tracks_per_day)


day
Friday       21840
Monday       21354
Wednesday    18059
Name: user_id, dtype: int64


El viernes tiene la mayor cantidad de reproducciones (21,840), lo que sugiere que los usuarios son más activos a finales de la semana.
El lunes ocupa el segundo lugar con 21,354 reproducciones, lo que indica que también es un día con alta actividad.
El miércoles tiene la menor cantidad de reproducciones (18,059), lo que sugiere que a mitad de semana la actividad de los usuarios disminuye.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [37]:
# Declarar la función number_tracks()
def number_tracks(df, day, city):
    """
    Filtra los datos por día y ciudad y cuenta el número de reproducciones.
    
    Parámetros:
    df (DataFrame): El DataFrame con los datos.
    day (str): Día de la semana a filtrar.
    city (str): Ciudad a filtrar.
    
    Retorna:
    int: Número de reproducciones en la ciudad y el día especificados.
    """
    # Filtrar datos por día
    filtered_data = df[df["day"] == day]
    
    # Filtrar datos por ciudad
    filtered_data = filtered_data[filtered_data["city"] == city]
    
    # Contar las reproducciones (número de valores en 'user_id')
    return filtered_data["user_id"].count()


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [39]:
# El número de canciones reproducidas en Springfield el lunes
# Listas de ciudades y días a analizar
cities = ["Springfield"]
days = ["Monday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Springfield el Monday: 15740


In [40]:
# El número de canciones reproducidas en Shelbyville el lunes
# Listas de ciudades y días a analizar
cities = ["Shelbyville"]
days = ["Monday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Shelbyville el Monday: 5614


In [41]:
# El número de canciones reproducidas en Springfield el miércoles
# Listas de ciudades y días a analizar
cities = ["Springfield"]
days = ["Wednesday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Springfield el Wednesday: 11056


In [42]:
# El número de canciones reproducidas en Shelbyville el miércoles
# Listas de ciudades y días a analizar
cities = ["Shelbyville"]
days = ["Wednesday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Shelbyville el Wednesday: 7003


In [43]:
# El número de canciones reproducidas en Springfield el viernes
# Listas de ciudades y días a analizar
cities = ["Springfield"]
days = ["Friday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Springfield el Friday: 15945


In [44]:
# El número de canciones reproducidas en Shelbyville el viernes
# Listas de ciudades y días a analizar
cities = ["Shelbyville"]
days = ["Friday"]

# Llamar a la función para cada combinación de ciudad y día
for city in cities:
    for day in days:
        count = number_tracks(df, day, city)
        print(f"Número de canciones reproducidas en {city} el {day}: {count}")


Número de canciones reproducidas en Shelbyville el Friday: 5895


**Conclusiones**

La hipótesis se confirma, ya que los datos muestran una clara diferencia en el consumo de música entre las dos ciudades.
Springfield tiene una mayor actividad musical en comparación con Shelbyville en los tres días analizados.
Además, la variación en el número de reproducciones según el día de la semana indica patrones de consumo diferentes, con mayor actividad los viernes y menor los miércoles.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

Los usuarios y usuarias de Springfield reproducen más música que los de Shelbyville, independientemente del día. Además, la actividad varía a lo largo de la semana, con una disminución los miércoles y un pico los viernes.
Estos resultados pueden ser útiles para estrategias de contenido y marketing, priorizando días de alta actividad en cada ciudad.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)